modules/table/src/main/java/org/apache/ignite/internal/table/SynchronousPriorityQueue.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.table;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.function.ToLongFunction;
+
+/**
+ * A thread-safe wrapper over {@link java.util.PriorityQueue}, which uses {@code long} value for ordering.
+ * The implementation provides a method to poll top item up to the given priority.
+ *
+ * @param <T> Item type.
+ */
+public class SynchronousPriorityQueue<T> {
+    /** A queue. Guarded by itself. */
+    private final PriorityQueue<T> queue;
+    private final ToLongFunction<T> priorityExtractor;
+
+    /**
+     * Creates a queue.
+     *
+     * @param priorityExtractor Priority extractor.
+     */
+    public SynchronousPriorityQueue(ToLongFunction<T> priorityExtractor) {
```

Comment:
It has become better, but it is still not obvious why there should be a `long`. 
Why am I grumbling, the class looks like a general one, but on the other hand it is very specific.

If we want to make it general, then I think it’s better to pass a function to retrieve the priority being compared (`int`, `long`, `string`, etc.). Or rename it to queue `LwmDestuctionQueue`.

Reply:
> `LwmDestuctionQueue` 

It does not used for "destroying LWM". LWM (a timestamp) is not used as a key.
Maybe, `LongPriorityQueue`?

Reply:
Still don't like the name.
It's up to you.